File: src/main/java/org/distril/beengine/world/chunk/SubChunk.java
package org.distril.beengine.world.chunk;

import io.netty.buffer.ByteBuf;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.distril.beengine.material.block.BlockState;

import java.util.Arrays;
import java.util.Objects;

@Getter
@RequiredArgsConstructor
public class SubChunk {

	public static final int VERSION = 8;
	public static final int COUNT = 16;

	private final Layer[] layers = {new Layer(), new Layer()};

	private final int index;


	public void setBlockState(int x, int y, int z, int layer, BlockState state) {
		this.layers[layer].set(x & 0xf, y & 0xf, z & 0xf, state);
	}

	public BlockState getBlockState(int x, int y, int z, int layer) {
		return this.layers[layer].get(x & 0xf, y & 0xf, z & 0xf);
	}

	public synchronized void writeToNetwork(ByteBuf buffer) {
		buffer.writeByte(VERSION);
		buffer.writeByte(this.layers.length);
		// buffer.writeByte(this.index);

		Arrays.stream(this.layers).forEach(layer -> layer.writeToNetwork(buffer));
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (obj == null || this.getClass() != obj.getClass()) {
			return false;
		}

		SubChunk that = (SubChunk) obj;
		return this.index == that.getIndex() && Arrays.equals(this.layers, that.getLayers());
	}

	@Override
	public int hashCode() {
		return Objects.hash(this.index, Arrays.hashCode(layers));
	}
}

File: src/main/kotlin/org/distril/beengine/world/chunk/SubChunk.kt
package org.distril.beengine.world.chunk

import io.netty.buffer.ByteBuf
import org.distril.beengine.material.block.BlockState
import java.util.*

class SubChunk(val index: Int) {

	val layers = arrayOf(Layer(), Layer())

	fun setBlockState(x: Int, y: Int, z: Int, state: BlockState, layer: Int = 0) {
		synchronized(this.layers) {
			this.layers[layer][x and 0xf, y and 0xf, z and 0xf] = state
		}
	}

	fun getBlockState(x: Int, y: Int, z: Int, layer: Int = 0) = this.layers[layer][x and 0xf, y and 0xf, z and 0xf]

	fun writeToNetwork(buffer: ByteBuf) {
		buffer.writeByte(VERSION)
		buffer.writeByte(this.layers.size)
		buffer.writeByte(this.index)

		synchronized(this.layers) {
			this.layers.forEach { it.writeToNetwork(buffer) }
		}
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is SubChunk) return false

		return this.index == other.index && this.layers.contentEquals(other.layers)
	}

	override fun hashCode() = Objects.hash(this.index, this.layers.contentHashCode())

	companion object {

		const val VERSION = 9

		const val COUNT = 16
	}
}
